Mathlib/Order/RelSeries.lean

```diff
@@ -0,0 +1,198 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.Logic.Equiv.Fin
+import Mathlib.Data.List.Indexes
+import Mathlib.Data.Rel
+
+/-!
+# Series of a relation
+
+If `r` is a relation on `α` then a relation series of length `n` is a series
+`a_0, a_1, ..., a_n` such that `r a_i a_{i+1}` for all `i < n`
+
+-/
+
+variable {α : Type _} (r : Rel α α)
+
+/--
+Let `r` be a relation on `α`, a relation series of `r` of length `n` is a series
+`a_0, a_1, ..., a_n` such that `r a_i a_{i+1}` for all `i < n`
+-/
+structure RelSeries where
+/-- the number of inequalities in the series -/
+length : ℕ
+/-- the underlying function of a relation series -/
+toFun : Fin (length + 1) → α
+/-- adjacent elements are related by the said relation -/
+step : ∀ (i : Fin length), r (toFun <| Fin.castSucc i) <| toFun <| i.succ
+
+namespace RelSeries
+
+instance : CoeFun (RelSeries r) (fun x ↦ Fin (x.length + 1) → α) :=
+{ coe := RelSeries.toFun }
+
+instance : Preorder (RelSeries r) :=
+  Preorder.lift fun x => x.length
+
+lemma le_def (x y : RelSeries r) : x ≤ y ↔ x.length ≤ y.length :=
+  Iff.rfl
+
+lemma lt_def (x y : RelSeries r) : x < y ↔ x.length < y.length :=
+  Iff.rfl
+
+/--
+For any type `α`, each term of `α` gives a relation series with the right most index to be 0.
+-/
+@[simps!] def singleton (a : α) : RelSeries r where
+  length := 0
+  toFun := fun _ => a
+  step := fun i => Fin.elim0 i
+
+instance [IsEmpty α] : IsEmpty (RelSeries r) where
+  false := fun x ↦ IsEmpty.false (x 0)
+
+instance [Inhabited α] : Inhabited (RelSeries r) where
+  default := singleton r default
+
+instance [Nonempty α] : Nonempty (RelSeries r) :=
+  Nonempty.map (singleton r) inferInstance
+
+variable {r}
+
+@[ext]
+lemma ext {x y : RelSeries r} (length_eq : x.length = y.length)
+    (toFun_eq : x.toFun = y.toFun ∘ Fin.cast (by rw [length_eq])) : x = y := by
+  rcases x with ⟨nx, fx⟩
+  rcases y with ⟨ny, fy⟩
+  dsimp at length_eq toFun_eq
+  subst length_eq
+  rw [Fin.cast_refl, OrderIso.coe_refl, Function.comp.right_id] at toFun_eq
+  subst toFun_eq
+  rfl
+
+lemma rel_of_lt [IsTrans α r] (x : RelSeries r) {i j : Fin (x.length + 1)} (h : i < j) :
+    r (x i) (x j) := by
+  induction i using Fin.inductionOn generalizing j with
+  | h0 => induction j using Fin.inductionOn with
+    | h0 => cases lt_irrefl _ h
+    | hs j ihj =>
+      by_cases H : 0 < Fin.castSucc j
+      . exact IsTrans.trans _ _ _ (ihj H) (x.step _)
+      . convert x.step _
+        simp only [not_lt, Fin.le_zero_iff] at H
+        exact H.symm
```

Comment:
```suggestion
      . simp only [not_lt, Fin.le_zero_iff] at H
        rw [← H]
        exact x.step _
```
I've been a bit edgy about non-terminal `convert`s ever since I had a discussion with Gabriel about them. This avoids the non-terminal `convert` without being any longer.